### ISSUE TITLE
CanEqual is undefined by default for unions?

### ISSUE THREAD

**Issue by soronpo** [lampepfl/dotty#19193]
## Compiler version

v3.3.1

## Minimized code

https://scastie.scala-lang.org/94w48wQbRdq9rQ2IKcbNCg
```Scala
import language.strictEquality
sealed trait Foo derives CanEqual
sealed trait Bar derives CanEqual

object FooObj extends Foo
object BarObj extends Bar

val x: Foo | Bar = FooObj

@main def main: Unit =
  x match   
    case FooObj =>
    case BarObj =>
```

## Output

```scala
error:  Values of types object FooObj and Foo | Bar cannot be compared with == or !=
```

## Expectation
I would expect that by default we have something like
```scala
given CanEqual[A, B](using CanEqual[A, A], CanEqual[B, B]): CanEqual[A | B] = CanEqual.derived
```
So no error should happen under strict equality for any union of types that have `CanEqual` defined for them.

**Comment by sjrd** [comment]
I don't think that would be a sound generalization. The fact that `A` can compare to `A` and `B` can compare to `B` does not give the ability for `A` to compare to `B`. But that's what a `CanEqual[A | B, A | B]` would promise.

For a concrete example, consider:
```scala
object Opaques:
  opaque type A = Int
  given CanEqual[A, A] = ...

  opaque type B = Int
  given CanEqual[B, B] = ...
```
I can safely compare an `A` to an `A`, and a `B` to a `B`. But I shouldn't be able to compare an `A` to a `B`, and hence a `CanEqual[A | B, A | B]` would not be desirable.

**Comment by soronpo** [comment]
Even with this addition this won't compile
```scala
given CanEqual[Foo, Bar] = CanEqual.derived
given CanEqual[Bar, Foo] = CanEqual.derived
```

**Comment by sjrd** [comment]
Sure, but that's starting to look a bit specific, no? When searching for `CanEqual[A | B, A | B]`, I don't think we want the compiler to look for 4 sub implicit searches. It grows exponentially with the size of the union too.

**Comment by mbovel** [comment]
Reproduced:

```scala
package unionEqual
import language.strictEquality
class Foo derives CanEqual
class Bar derives CanEqual

@main def main: Unit =
    val x: Foo | Bar = ???
    val y: Foo | Bar = ???
    x == y // error: Values of types unionEqual.Foo | unionEqual.Bar and unionEqual.Foo | unionEqual.Bar cannot be compared with == or !=
``` 

Adding `given CanEqual[Foo | Bar, Foo | Bar] = CanEqual.derived` works. Does that work for your use-case @soronpo?

**Comment by soronpo** [comment]
> Adding `given CanEqual[Foo | Bar, Foo | Bar] = CanEqual.derived`

This is the trivial solution that will always work.

**Comment by odersky** [comment]
I also think there's nothing we can do on the compiler side, unless someone has a new approach how to handle this,

**Comment by Sporarum** [comment]
Would there be a way to allow only comparisons between `A` and `A | B`, but not between `A` and `B` ?

It seems very constraining to have a variable that's "A or something else" and not being able to test if it is "This specific instance of A"

**Comment by som-snytt** [comment]
I would expect patmat to downcast the scrutinee. (If multiple `CanEqual` are available, it can pick one.)
```
@main def main: Unit = println:
  x match
  case y: Foo if FooObj == y => 0 // case FooObj
  case y: Bar if BarObj == y => 1
```
but with exhaustivity unimpaired.

`eq` and `case _: FooObj.type` suffice for instance test.

**Comment by Sporarum** [comment]
> I also think there's nothing we can do on the compiler side, unless someone has a new approach how to handle this,

I don't know if it is a good one, but:

https://contributors.scala-lang.org/t/can-we-improve-canequal/7369/2
